tide/core/utils/IterableSmartPtrCollection.h

```diff
@@ -0,0 +1,131 @@
+/* or implied, of Ecole polytechnique federale de Lausanne.          */
+/*********************************************************************/
+
+#ifndef ITERABLE_SMART_PTR_COLLECTION_H
+#define ITERABLE_SMART_PTR_COLLECTION_H
+
+#include <iterator>
+#include <type_traits>
+
+/**
+ * Facade class that lets users safely iterate over a collection of smart_ptrs.
+ *
+ * Iterating a collection through this class gives direct access to the pointed
+ * objects, helping:
+ * a) Enforce const correctness: the pointed objects can't be mutated when
+ *    iterating over a const container<smart_ptr>.
+ * b) Improve readability: no need to dereference the pointer in user code.
+ */
+template <typename ContainerT>
+class IterableSmartPtrCollection
+{
+public:
+    template <bool is_const_iterator = true>
+    struct smart_ptr_iterator_wrapper
+    {
+        // std::iterator_traits
+        using difference_type = typename ContainerT::iterator::difference_type;
+        using value_type = typename ContainerT::value_type::element_type;
+        using reference =
+            typename std::conditional_t<is_const_iterator, const value_type&,
+                                        value_type&>;
+        using pointer =
+            typename std::conditional_t<is_const_iterator, const value_type*,
+                                        value_type*>;
+        using iterator_category = std::forward_iterator_tag;
+
+        // internal
+        using smart_ptr_iterator =
+            typename std::conditional_t<is_const_iterator,
+                                        typename ContainerT::const_iterator,
+                                        typename ContainerT::iterator>;
+
+        smart_ptr_iterator_wrapper(smart_ptr_iterator it)
+            : _it{it}
+        {
+        }
+        smart_ptr_iterator_wrapper operator++()
+        {
+            return smart_ptr_iterator_wrapper(++_it);
+        }
+        bool operator!=(smart_ptr_iterator_wrapper other)
+        {
+            return _it != other._it;
+        }
+        reference operator*() { return *(*_it); }
+        pointer operator->() const { return _it->get(); }
+    private:
+        smart_ptr_iterator _it;
+    };
+
+    using iterator = smart_ptr_iterator_wrapper<false>;
+    using const_iterator = smart_ptr_iterator_wrapper<true>;
+    using value_type = typename ContainerT::value_type::element_type;
+
+    IterableSmartPtrCollection(ContainerT& container)
+        : _container(container)
+    {
+    }
+
+    template <typename T = ContainerT>
+    typename std::enable_if<!std::is_const<T>::value, iterator>::type begin()
+    {
+        return iterator{_container.begin()};
+    }
+
+    template <typename T = ContainerT>
+    typename std::enable_if<!std::is_const<T>::value, iterator>::type end()
+    {
+        return iterator{_container.end()};
+    }
+
+    auto begin() const { return const_iterator{_container.begin()}; }
```

Comment:
cbegin()/cend() C+11 style?

Reply:
yes, that's a good idea, I can add that. I think begin() const is still needed for range-based for though, so we need both.